test/run_test.py

```diff
@@ -0,0 +1,152 @@
+#!/usr/bin/env python
+
+import argparse
+import os
+import subprocess
+import sys
+
+TESTS = [
+    'autograd',
+    'cpp_extensions',
+    'cuda',
+    'dataloader',
+    'distributed',
+    'distributions',
+    'indexing',
+    'jit',
+    'legacy_nn',
+    'multiprocessing',
+    'nccl',
+    'nn',
+    'optim',
+    'sparse',
+    'torch',
+    'utils',
+]
+
+WINDOWS_BLACKLIST = [
+    'cpp_extensions',
+    'distributed',
+]
+
+
+def shell(command, cwd):
+    popen = subprocess.Popen(
+        command,
+        stdout=subprocess.PIPE,
+        universal_newlines=True,
+        cwd=cwd,
+        shell=True)
+    for stdout_line in iter(popen.stdout.readline, ''):
+        print(stdout_line.strip())
+    popen.stdout.close()
+    return_code = popen.wait()
+    return return_code == 0
+
+
+def get_shell_output(command):
+    return subprocess.check_output(command, shell=True).decode().strip()
+
+
+def run_test(python, test_module, cwd, verbose):
+    verbose = '--verbose' if verbose else ''
+    shell('{} {} {}'.format(python, test_module, verbose), cwd)
+
+
+def test_cpp_extensions(python, test_module, test_directory, verbose):
+    shell('{} setup.py install --root ./install'.format(python),
+          os.path.join(test_directory, 'cpp_extensions'))
+
+    python_path = os.environ.get('PYTHONPATH', '')
+    install_directory = get_shell_output(
+        "find cpp_extensions/install -name '*-packages'")
+    install_directory = os.path.join(test_directory, install_directory)
+    os.environ['PYTHONPATH'] = '{}:{}'.format(install_directory, python_path)
+    run_test(python, test_module, test_directory, verbose)
+    os.environ['PYTHONPATH'] = python_path
+
+
+def test_distributed(python, test_module, cwd, verbose):
+    os.environ['PYCMD'] = python
+    shell('./run_distributed_tests.sh', cwd)
+
+
+CUSTOM_HANDLERS = {
+    'cpp_extensions': test_cpp_extensions,
+    'distributed': test_distributed,
+}
+
+
+def parse_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('-v', '--verbose', action='store_true')
+    parser.add_argument('-p', '--python')
+    parser.add_argument('-c', '--coverage', action='store_true')
+    parser.add_argument(
+        '-i', '--include', nargs='+', choices=TESTS, default=TESTS)
+    parser.add_argument(
+        '-x', '--exclude', nargs='+', choices=TESTS, default=[])
+    parser.add_argument('-f', '--first', choices=TESTS)
+    parser.add_argument('-l', '--last', choices=TESTS)
+    parser.add_argument('--with-windows', action='store_true')
+    parser.add_argument('--test-dir')
+    return parser.parse_args()
+
+
+def get_python_command(options):
+    if options.coverage:
+        return 'coverage run --parallel-mode --source torch'
+    elif options.python:
+        return options.python
+    else:
+        return os.environ.get('PYCMD', 'python')
+
+
+def get_selected_tests(options):
+    selected_tests = list(sorted(set(options.include) - set(options.exclude)))
+
+    if options.first:
```

Comment:
This is a bit weird, because it's not running tests between `first` and `last` in the order of `TESTS`, but in the alphabetical order (because of `sorted` above). It would be better to just take `options.include` and use `.remove` to erase everything from `.exclude`. It is slower, but you're running it on 20 elements.